### PR TITLE
Add test for #154

### PR DESCRIPTION
**Pull request by siqpush**
Guard against duplicate source registrations

### CODE REVIEW

**Comment by notgull** [comment]
You need to fix the Windows tests by using `&socket1` instead of `socket1.borrow().as_fd()`.

File: tests/async.rs [hunk 1]
```diff
@@ -356,3 +359,41 @@ fn shutdown() -> io::Result<()> {
         Ok(())
     })
 }
+
+// prevent source from unregistering by trying to register it twice
+#[test]
+fn duplicate_socket_insert() -> io::Result<()> {
+    future::block_on(async {
+        let (socket1, socket2) = match Async::<UnixStream>::pair() {
```
Comment (notgull) [inline tests/async.rs:367]: Instead of this, use the `tcp_pair` function in this module.

File: tests/async.rs [hunk 2]
```diff
@@ -356,3 +356,33 @@ fn shutdown() -> io::Result<()> {
...
+// prevent source from unregistering by trying to register it twice
+#[test]
+fn duplicate_socket_insert() -> io::Result<()> {
+    future::block_on(async {
+        let listener = Async::<TcpListener>::bind(([127, 0, 0, 1], 0))?;
+        let addr = listener.as_ref().local_addr()?;
+
+        // attempt to register twice
+        assert!(
+            Async::<TcpListener>::bind(addr).is_err(),
```
Comment (notgull) [inline tests/async.rs:369]:
```suggestion
            Async::<TcpListener>::new(&listener).is_err(),
```

Comment (siqpush) [reply tests/async.rs:369]: @notgull only additional change to suggestion is removing the type annotation

**Review by notgull — APPROVED** [review]
Thanks!

**Pull request merged**